src/Connection/DefaultConnectionPool.php

```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace Amp\Http\Client\Connection;
+
+use Amp\Http\Client\Request;
+use Amp\Socket\Socket;
+
+final class DefaultConnectionPool implements ConnectionPool
+{
+    private $connections = [];
+
+    public function createConnection(Socket $socket): Connection
+    {
+        $connection = new Http1Connection($socket);
+        $this->connections[$socket->getRemoteAddress()->toString()] = $connection;
```

Comment:
I wanted to say this only allows one concurrent connection per host, but actually this works, because it has the port in it.

Reply:
So it allows one concurrent connection per host per port, which isn't really enough for HTTP/1.x, but works fine for HTTP/2. This is one of the areas where I'm not exactly sure how I want to approach the design.

Reply:
Ugh, well, right. I think I'd add `Connection::isBusy` and `ConnectionPool` chooses the next connection then or creates a new one if none is available. This also allows limiting the number of requests per HTTP/2 connection then, so the connection manager creates a separate HTTP/2 connection at some point.

The major issue here is that we can't use the remote address as key, because different sites might be hosted on the same IP, but due to SNI there might be different TLS certificates served.

Reply:
Ugh, well, right. I think I'd add `Connection::isBusy` and `ConnectionPool` chooses the next connection then or creates a new one if none is available. This also allows limiting the number of requests per HTTP/2 connection then, so the connection manager creates a separate HTTP/2 connection at some point.

The major issue here is that we can't use the remote address as key, because different sites might be hosted on the same IP, but due to SNI there might be different TLS certificates served.